switch to glide

deads2k opened the pull request.

deads2k commented:
@liggitt alright, this builds locally for me and it creates symlinks for the staging folders when you do `hack/update-deps.sh`

liggitt commented:
> it creates symlinks for the staging folders when you do hack/update-deps.sh

running that left a diff in my working dir... that expected?
```
$ git diff --numstat
4       2       glide.lock
0       2       vendor/github.com/getsentry/raven-go/docs/_sentryext/.gitignore
0       13      vendor/github.com/getsentry/raven-go/docs/_sentryext/README.rst
0       708     vendor/github.com/getsentry/raven-go/docs/_sentryext/sentryext.py
0       113     vendor/github.com/getsentry/raven-go/docs/_sentryext/verify-docs.py
14      8       vendor/google.golang.org/appengine/.travis.yml
26      26      vendor/google.golang.org/appengine/README.md
4       0       vendor/google.golang.org/appengine/aetest/instance.go
3       0       vendor/google.golang.org/appengine/aetest/instance_test.go
23      17      vendor/google.golang.org/appengine/aetest/instance_vm.go
37      0       vendor/google.golang.org/appengine/appengine.go
1       37      vendor/google.golang.org/appengine/appengine_vm.go
1       1       vendor/google.golang.org/appengine/capability/capability.go
4       0       vendor/google.golang.org/appengine/channel/channel.go
1       1       vendor/google.golang.org/appengine/cloudsql/cloudsql.go
1       1       vendor/google.golang.org/appengine/cloudsql/cloudsql_vm.go
2       2       vendor/google.golang.org/appengine/cmd/aebundler/aebundler.go
24      216     vendor/google.golang.org/appengine/cmd/aedeploy/aedeploy.go
3       2       vendor/google.golang.org/appengine/datastore/datastore.go
277     32      vendor/google.golang.org/appengine/datastore/datastore_test.go
26      16      vendor/google.golang.org/appengine/datastore/doc.go
135     40      vendor/google.golang.org/appengine/datastore/load.go
4       5       vendor/google.golang.org/appengine/datastore/metadata.go
81      45      vendor/google.golang.org/appengine/datastore/prop.go
99      156     vendor/google.golang.org/appengine/datastore/prop_test.go
11      0       vendor/google.golang.org/appengine/datastore/query.go
46      19      vendor/google.golang.org/appengine/datastore/save.go
25      5       vendor/google.golang.org/appengine/delay/delay.go
106     0       vendor/google.golang.org/appengine/delay/delay_test.go
1       1       vendor/google.golang.org/appengine/demos/guestbook/app.yaml
1       1       vendor/google.golang.org/appengine/demos/guestbook/guestbook.go
1       1       vendor/google.golang.org/appengine/demos/helloworld/helloworld.go
2       1       vendor/google.golang.org/appengine/internal/aetesting/fake.go
46      9       vendor/google.golang.org/appengine/internal/api.go
35      3       vendor/google.golang.org/appengine/internal/api_classic.go
35      13      vendor/google.golang.org/appengine/internal/api_common.go
24      25      vendor/google.golang.org/appengine/internal/api_test.go
39      9       vendor/google.golang.org/appengine/internal/identity_classic.go
5       1       vendor/google.golang.org/appengine/internal/identity_vm.go
0       34      vendor/google.golang.org/appengine/internal/internal.go
0       58      vendor/google.golang.org/appengine/internal/internal_test.go
580     219     vendor/google.golang.org/appengine/internal/search/search.pb.go
6       0       vendor/google.golang.org/appengine/internal/search/search.proto
2       2       vendor/google.golang.org/appengine/log/log.go
1       1       vendor/google.golang.org/appengine/mail/mail.go
19      15      vendor/google.golang.org/appengine/memcache/memcache_test.go
0       52      vendor/google.golang.org/appengine/namespace_test.go
40      20      vendor/google.golang.org/appengine/remote_api/client.go
19      0       vendor/google.golang.org/appengine/remote_api/client_test.go
1       1       vendor/google.golang.org/appengine/remote_api/remote_api.go
2       2       vendor/google.golang.org/appengine/runtime/runtime.go
8       4       vendor/google.golang.org/appengine/search/doc.go
108     28      vendor/google.golang.org/appengine/search/search.go
378     14      vendor/google.golang.org/appengine/search/search_test.go
12      6       vendor/google.golang.org/appengine/search/struct.go
43      2       vendor/google.golang.org/appengine/search/struct_test.go
1       1       vendor/google.golang.org/appengine/socket/doc.go
2       2       vendor/google.golang.org/appengine/socket/socket_vm.go
45      0       vendor/google.golang.org/appengine/taskqueue/taskqueue.go
57      0       vendor/google.golang.org/appengine/taskqueue/taskqueue_test.go
11      2       vendor/google.golang.org/appengine/user/user_classic.go
```

mfojtik commented on hack/update-deps.sh:
```diff
@@ -0,0 +1,44 @@
+for pkg in vendor/k8s.io/kubernetes/staging/src/k8s.io/*; do
+  dir=$(basename $pkg)
+  unlink vendor/k8s.io/$dir || true
+  rm -rf vendor/k8s.io/$dir || true
```
@deads2k will this leave the deps broken when the glide below fails?

deads2k replied:
> @deads2k will this leave the deps broken when the glide below fails?

Yeah.  We'll call that something to iterate on.

mfojtik commented on glide.yaml:
```diff
@@ -0,0 +1,69 @@
+- package: k8s.io/kubernetes
+  repo:    git@github.com:openshift/kubernetes
+  version: release-1.8.1
+# only used by pkg/build/builder
```
i thought pkg/build/builder is excluded?

deads2k replied:
> i thought pkg/build/builder is excluded?

they have an incomplete vendor folder.  Go them.  Another thing to fix up post-this.

mfojtik commented on glide.yaml:
```diff
@@ -0,0 +1,69 @@
+  version: release-1.8.1
+- package: k8s.io/kubernetes
+  repo:    git@github.com:openshift/kubernetes
+  version: release-1.8.1
```
is this right or it should be rebase-1.8.1 ?

mfojtik commented on glide.yaml:
```diff
@@ -0,0 +1,69 @@
+  version: a8ee86b1cce0c13bd541a99140682a92635ba9f7
+- package: github.com/containers/image
+  repo:    git@github.com:openshift/containers-image
+  version: openshift-3.8
```
so does this mean that if I want to bump this I will have to bump the fork branch and then re-run update-deps ?

deads2k replied:
> so does this mean that if I want to bump this I will have to bump the fork branch and then re-run update-deps ?

Yeah. Or you can find the person who made that ridiculous carry and make them fix it.

mfojtik commented:
@deads2k for me it exploded even before I got the diff:

```
[INFO]	--> Setting version for github.com/seccomp/libseccomp-golang to 1b506fc7c24eec5a3693cdcbed40d9c226cfc6a1.
[ERROR]	Failed to retrieve a list of dependencies: Error resolving imports
[ERROR] PID 6262: hack/update-deps.sh:37: `glide update --strip-vendor` exited with status 1.
[INFO] 		Stack Trace:
[INFO] 		  1: hack/update-deps.sh:37: `glide update --strip-vendor`
[INFO]   Exiting with code 1.
```

mfojtik commented:
@deads2k @liggitt this fixed the appengine diff: https://github.com/deads2k/origin/commit/e9ee6ff06b0d006c0e82289723386e90ee76d9c1

tnozicka commented on hack/update-deps.sh:
```diff
@@ -0,0 +1,44 @@
+
+source "$(dirname "${BASH_SOURCE}")/lib/init.sh"
+
+# fail early if any of the staging dirs is checked out
```
why do we care about this in update script?

mfojtik replied:
symlinks

tnozicka replied:
how is GOPATH relevant to symlinks which are in vendor folder? vendor takes precedence

deads2k replied:
> how is GOPATH relevant to symlinks which are in vendor folder? vendor takes precedence

We symlink into vendor. This is pre-existing and it looks correct to me.

tnozicka replied:
to me it seems unrelated to the task. If the bellow lines are not affected by stuff present in GOPATH why does it check for whats there? If this is preexisting, that's likely because of Godeps specifics when it worked with your GOPATH. Can't see any relations to glide.

sttts replied:
We used to have hard to debug re-vendoring errors where the GOPATH version was picked up and the result differed depending on the developer's GOPATH. This was added as a sanity check to avoid this situation.

tnozicka commented on hack/update-deps.sh:
```diff
@@ -0,0 +1,44 @@
+#  "github.com/jteeuwen/go-bindata/go-bindata"
+#)
+
+# remove symlinks for glide update
```
glide switches vendor/ and tmp/vendor/ as far as I recall so this shouldn't be needed or am I wrong?

tnozicka commented on hack/update-deps.sh:
```diff
@@ -0,0 +1,44 @@
+  dir=$(basename $pkg)
+  rm -rf vendor/k8s.io/$dir
+  ln -s kubernetes/staging/src/k8s.io/$dir vendor/k8s.io/$dir
+done
```
\n

tnozicka commented on hack/update-deps.sh:
```diff
@@ -0,0 +1,44 @@
+for pkg in vendor/k8s.io/kubernetes/staging/src/k8s.io/*; do
+  dir=$(basename $pkg)
+  rm -rf vendor/k8s.io/$dir
+  ln -s kubernetes/staging/src/k8s.io/$dir vendor/k8s.io/$dir
```
I think `ln` can do this in one step, it was something like `ln -sfn`

deads2k replied:
> I think ln can do this in one step, it was something like ln -sfn

tried, does not work

tnozicka commented on glide.yaml:
```diff
@@ -0,0 +1,69 @@
+  repo:    git@github.com:openshift/containers-image
+  version: openshift-3.8
+- package: github.com/vjeantet/ldapserver
+  version: v1.0
```
\n

tnozicka commented:
+100 for switching to glide.

Since I have been there I don't think the symlinking can ever work properly. The issue is that glide has to resolve all the dependencies in `glide update` step and if there are some files missing, it fails. So you need have the publishing repository updated first. Putting your faith into probability you could hope that the changes were so small and not changing dependency resolution that you will be fine but even a simple file addition/move/backport in upstream will cause it to fail. But since that establishes the need for the repository being published upfront you can include that repository directly and there is no point for symlinking.

The appropriate staging repo commits for kubernetes commit `f6401` can be found by
```
REPO=client-go && \
git --git-dir=${GOPATH}/src/k8s.io/${REPO}/.git --work-tree=${GOPATH}/src/k8s.io/${REPO} log --all --follow -- kubernetes-sha | grep -B4 f6401
```

soltysh commented:
The only worrying thing for me is the huge diff here, but I assume it's godep vs glide thing, I'll be looking more into it after tomorrow.

mfojtik commented:
@soltysh the diff is due to appengine version in pkg/build/builder I have commit that pin the version to the sha256 used in our level, so we should be fine?

deads2k commented:
> running that left a diff in my working dir... that expected?

Actually yes.  I didn't explicitly pin levels.  We'll have to decide whether we're really going to pin everything, whether we'll try to find branches, whether we'll try to match kube (which I think happens by inspection of their godep.json), or whether we'll try to stay current when possible.

Pinning levels is heavy, but possible if we want to.  I did go ahead and pin that one.

deads2k commented:
@liggitt @mfojtik alright, see if you can live with what I've done.  I disabled some tests we weren't running before, removed the oadm completions a bit early (they looked like the only ones that were broken), and updated some openapi generation excludes.

deads2k commented:
@mfojtik looks like I'm one unit test away.  Take a look in your AM?

liggitt commented:
Need the other half of the upstream PR that changed that unit test

xiaods commented:
need rebase

deads2k commented:
> Need the other half of the upstream PR that changed that unit test

Yeah, wasn't messing with at 9pm though.  What were *you* doing?....

deads2k commented:
Alright, this adds

```
+    flags+=("--log-flush-frequency=")
+    flags+=("--v=")
+    flags+=("--vmodule=")
```

to completions, but it makes the re-rebase, the openshift/api update, and the 1.9 rebase so much easier that I think we should take it as-is and see what we can do about those later.

sttts commented on glide.yaml:
```diff
@@ -0,0 +1,76 @@
+- package: github.com/emicklei/go-restful-swagger12
+  repo:    https://github.com/openshift/emicklei-go-restful-swagger12
+  version: release-1.0.1
+- package: google.golang.org/appengine
```
that are more than I expected. It's the minimal set? Would be great to have a comment with the reason why we chose another version than upstream.

deads2k replied:
> that are more than I expected. It's the minimal set? Would be great to have a comment with the reason why we chose another version than upstream.

Explanations as a followup as we try to prune.  This is the minimal set for now.

sttts commented on hack/godep-restore.sh:
```diff
@@ -1,85 +0,0 @@
-#!/bin/bash
-
-source "$(dirname "${BASH_SOURCE}")/lib/init.sh"
```
-258 +37 lines 👍 👍 👍

sttts commented on hack/update-deps.sh:
```diff
@@ -0,0 +1,37 @@
+# recreate symlinks after vendoring
+for pkg in vendor/k8s.io/kubernetes/staging/src/k8s.io/*; do
+  dir=$(basename $pkg)
+  rm -rf vendor/k8s.io/$dir
```
can you add a `cd "${BASH_SOURCE}/.."` at the top? Otherwise this relative `rm -rf` is concerning.

sttts replied:
And as we only delete symlinks, a simple `rm -f` (or even `rm`?) should be enough.

deads2k replied:
> And as we only delete symlinks, a simple rm -f (or even rm?) should be enough.

It's for deleting the directories if they exist from a bad run previously.  I had some when I ran.

mfojtik commented:
/lgtm

RIP godeps.

sttts commented on hack/update-deps.sh:
```diff
@@ -0,0 +1,37 @@
+# Some things we want in godeps aren't code dependencies, so ./...
+# won't pick them up.
+# TODO seems like this should be failing something somewhere
+#REQUIRED_BINS=(
```
```
rm -rf vendor/k8s.io/code-generator &&
git clone --branch master git@github.com:kubernetes/code-generator.git vendor/k8s.io/code-generator &&
pushd vendor/k8s.io/code-generator &>/dev/null &&
git checkout <SHA> &&
rm -rf .git &&
popd >/dev/null
```

liggitt commented on vendor/k8s.io/kubernetes/cmd/kube-apiserver/app/testing/server_test.go:
```diff
@@ -152,7 +152,7 @@ func TestCRDShadowGroup(t *testing.T) {
 }
 
-func TestCRD(t *testing.T) {
+func ETestCRD(t *testing.T) {
```
?

deads2k replied:
> ?

Disables a test (when I thought I had one to two) without making a large diff to deal with.  I think its a lack of entropy creating new certs problem.

deads2k commented:
/retest

deads2k commented:
/retest
